fix (forms): clear selected options when model is not an array

gary-b opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
1. For a select[multiple] with NgModel, the expected model type is an Array. Currently if you set it to a different type of object (ignoring null/undefined for the moment), you get an error. NgModel does not throw such an error when you use an invalid model value with other input controls, eg input[type=number].
   The repro in Issue #11926 includes the use of ngModel on a select element, where ngModel is not explicitly bound. This result in an initial model value for the select of an empty string. (I'm going to raise a separate issue about this ngModel behavior, which I suspect is related to one time string initialization.)
2. When you set the model value to null or undefined, the currently selected option elements stay selected. NgModel used with other input controls clears the selection when these values are passed in.

**What is the new behavior?**
1. If a select[multiple] is bound to an invalid value, the currently selected option elements are cleared and no error is raised.
2. If you set the model to null or undefined, the currently selected option elements are cleared.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No - (Well, changes what happens when the model is set to null / undefined - but I doubt anyone desires the old behavior.)
```

gary-b commented:
I fixed a lint error in my code. 2 of the travis-ci build steps are still failing though:
Fail 3 looks like infrastructure problems.
Fail 4 relates to a failing test in a completely different area.

vicb commented on modules/@angular/forms/src/directives/select_multiple_control_value_accessor.ts:
```diff
@@ -67,11 +67,16 @@ export class SelectMultipleControlValueAccessor implements ControlValueAccessor
-    this._optionMap.forEach((opt, o) => { opt._setSelected(ids.indexOf(o.toString()) > -1); });
+    let optionSelectedStateSetter: (opt: NgSelectMultipleOption, o: any) => void;
+    if (Array.isArray(value)) {
+      let values: Array<any> = <Array<any>>value;
```
`const values: any[] = <any[]>values;`

but using `value` directly should also work because of `Array.isArray(value)`

vicb commented on modules/@angular/forms/test/template_integration_spec.ts:
```diff
@@ -701,6 +701,66 @@ export function main() {
     });
 
+    describe('select multiple controls', () => {
+      var fixture: ComponentFixture<NgModelSelectMultipleForm>;
```
var -> let (or const where applicable)

kara commented on modules/@angular/forms/test/template_integration_spec.ts:
```diff
@@ -1079,6 +1139,19 @@ class NgModelSelectForm {
+  `
+})
+class NgModelSelectMultipleForm {
+  @Input() selectedCities: any[];
```
The @Input decorator here isn't necessary - can you remove?

kara commented on modules/@angular/forms/test/template_integration_spec.ts:
```diff
@@ -701,6 +701,66 @@ export function main() {
+        comp.cities = [{'name': 'SF'}, {'name': 'NYC'}, {'name': 'Buffalo'}];
+      });
+
+      const setSelectedCitiesDetectChangesAndTick = (selectedCities: any): void => {
```
Can you rename to `setSelectedCities`? The name is a bit of a handful to read.

kara commented on modules/@angular/forms/test/template_integration_spec.ts:
```diff
@@ -701,6 +701,66 @@ export function main() {
+      const assertOptionElementSelectedState = (selectedStates: boolean[]): void => {
+        const options = fixture.debugElement.queryAll(By.css('option'));
+        if (options.length !== selectedStates.length) {
+          throw `there are ${options.length} option elements but ${selectedStates.length}` +
```
I'd wrap the whole error message in backticks so you don't have to concat the string.

kara reviewed (commented):
Minor comments, but looks good

Commit:
fix (forms): clear selected options when model is not an array

When an invalid model value (eg empty string) was preset ngModel on
select[multiple] would throw an error, which is inconsistent with how it
works on other user input elements. Setting the model value to null or
undefined would also have no effect on what was already selected in the
UI. Fix this by clearing selected options when model set to null,
undefined or a type other than Array.

Closes #11926

gary-b commented:
PR updated.

kara reviewed (approved):
LGTM
